### ISSUE TITLE
RightCurly: remove obsolete shouldStartLine property

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#4022]
base on #3919, #3920 , #3921 , #3922 

Property `shouldStartLine` looks strange in design and conflicting with `option` property.
`option` property if main property that have bigger priority and clearly define behavior.

`shouldStartLine` was added at https://github.com/checkstyle/checkstyle/commit/78e9ce870170ca3ef88fd78f8e0af0681c958eb3#diff-778fdd60391f2ca28a68b504829bfe82
to cover cases of single line IF statements.

But to cover single line statements we introduced recently new mode `alone_or_singleline` for `option`  property at https://github.com/checkstyle/checkstyle/commit/f1dced161ab0c818f0e1e125fc5b88b71f5b171e

So property `shouldStartLine` should be removed from Check , 
if user used `shouldStartLine=false` so user need to use `option=alone_or_singleline` http://checkstyle.sourceforge.net/property_types.html#rcurly . 

It will be breaking compatibility update but this is ok, we have to do this.

**Comment by MEZk** [comment]
I'm on it

**Comment by MEZk** [comment]
@romani 

> if user used shouldStartLine=false so user need to use option=alone_or_singleline

Not exactly. As it turned out shouldStartLine is a part of logic which deals with [double brace collection initializers](https://github.com/MEZk/checkstyle/blob/b4e884c2ff3bef182b045692b59c1aceae3cb892/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSingleline.java#L119) and [anonymous single line classes](https://github.com/checkstyle/checkstyle/blob/b4e884c2ff3bef182b045692b59c1aceae3cb892/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSingleline.java#L111). Thus, if we want to remove the option it will require to change the logic. 

Now I came up with two approaches:
1) Remove the option and move the condition which involves [isOnStartOfLine](https://github.com/MEZk/checkstyle/blob/70ab9f4af3d163b940e025bc4665f2f23892525a/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L214) to [shouldBeAloneOnLine](https://github.com/MEZk/checkstyle/blob/70ab9f4af3d163b940e025bc4665f2f23892525a/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L238). So the method will become:
```java
private boolean shouldBeAloneOnLine(RightCurlyOption bracePolicy, Details details) {
    final String targetSourceLine = getLines()[details.rcurly.getLineNo() - 1];
    return bracePolicy == RightCurlyOption.ALONE
            && !isAloneOnLine(details)
            && !isEmptyBody(details.lcurly)
            || bracePolicy == RightCurlyOption.ALONE_OR_SINGLELINE
            && !isAloneOnLine(details)
            && !isSingleLineBlock(details)
            && !isAnonInnerClassInit(details.lcurly)
            && !isEmptyBody(details.lcurly);
            || !isOnStartOfLine(details, targetSourceLine);
}
```

That will work, however, we will need to reimplement shouldBeAloneOnLine in future as a separate issue as ```|| !isOnStartOfLine(details, targetSourceLine);``` at the end is strange and irrational as for me. In addition, [the implementation](https://github.com/checkstyle/checkstyle/blob/70ab9f4af3d163b940e025bc4665f2f23892525a/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java#L257-L257) of ```isOnStartOfLine``` is wierd.

2) Make all changes (removing of option and reimplementation of ```shouldBeAloneOnLine```) in the scope of this issue.

**Comment by romani** [comment]
@MEZk , sad.

>This will require many changes as now all methods which determines whether right curly is alone on the line does not treat double brace initializers and anonymous classes properly.

It does not sound as easy change, please split in separate issues requests to update logic to support " double brace initializers" and "anonymous classes" properly without usage of `shouldStartLine` field. 2 separate issues, to ease regression testing.

**Comment by MEZk** [comment]
@romani 
1) https://github.com/checkstyle/checkstyle/issues/4057 is for double brace initializers.
2) https://github.com/checkstyle/checkstyle/issues/4059 is for anonymous classes.

**Comment by MEZk** [comment]
This issue is blocked by #4057 and #4059.

**Comment by romani** [comment]
fix is merged.

@ps-sp , thanks a lot for your hard work.

**Comment by rnveach** [comment]
@romani Master is failing because of RightCurly Test.
https://travis-ci.org/checkstyle/checkstyle/jobs/478016204#L1328

**Comment by romani** [comment]
fixed.
